packages/component-base/src/data-provider-controller/cache.js

```diff
@@ -0,0 +1,187 @@
+export class Cache {
+  /**
+   * @type {import('../data-provider-controller.js').DataProviderController}
+   */
+  controller;
+
+  /**
+   * The number of items.
+   *
+   * @type {number}
+   */
+  size = 0;
+
+  /**
+   * The number of items to display per page.
+   *
+   * @type {number}
+   */
+  pageSize;
+
+  /**
+   * The total number of items, including items from expanded sub-caches.
+   *
+   * @type {number}
+   */
+  effectiveSize = 0;
+
+  /**
+   * An array of cached items.
+   *
+   * @type {object[]}
+   */
+  items = [];
+
+  /**
+   * A map where the key is a requested page and the value is a callback
+   * that will be called with data once the request is complete.
+   *
+   * @type {Map<number, Function>}
+   */
+  pendingRequests = new Map();
+
+  /**
+   * A map where the key is the index of an item in the `items` array
+   * and the value is a sub-cache associated with that item.
+   *
+   * Note, it's intentionally defined as an object instead of a Map
+   * to ensure that Object.entries() returns an array with keys sorted
+   * in alphabetical order, rather than the order they were added.
+   *
+   * @type {Record<number, Cache>}
+   */
+  #subCacheByIndex = {};
+
+  /**
+   * @param {Cache['controller']} controller
+   * @param {number} pageSize
+   * @param {number | undefined} size
+   * @param {Cache | undefined} parentCache
+   * @param {number | undefined} parentCacheIndex
+   */
+  constructor(controller, pageSize, size, parentCache, parentCacheIndex) {
+    this.controller = controller;
+    this.pageSize = pageSize;
+    this.size = size || 0;
+    this.effectiveSize = size || 0;
+    this.parentCache = parentCache;
+    this.parentCacheIndex = parentCacheIndex;
```

Comment:
**argument:** Previously, the cache class kept a direct reference to the `parentItem`. However, by passing `parentCacheIndex`, we can simplify this [place](https://github.com/vaadin/flow-components/pull/5239/files#diff-255f48ed31d114d33fd62d5f44b260202e622606b1f6ffb8fa39fbbc96306a84L229-L231) in the Grid connector.